book/src/rfc/const_generic.md

```diff
@@ -0,0 +1,265 @@
+
+### Introduce ComputedExpr
+
+With the const variable as array size, we need to propagate its value during circuit generation to check if the access is within the bounds of the array. So we need a way to retain structure information of a variable that is currently represented by `VarOrRef`.
+
+We can add a new layer to the `VarOrRef` to keep track of the structure of the data. It is called `ComputedExpr`, as it is computed from `compute_expr` function, which resolves and passing around the values (`cvars`) via the `Expr`. The key difference from `VarOrRef` is `ComputedExpr` retains the structure of the underlying variables.
+
+`ComputedExpr` enables the field / array access to determine the structure of the value with actual size of the targeted access. For example, when we are accessing `houses[1].rooms[2]`, it can check if the access is within the bounds of `houses` and `rooms`, by checking the structure and size of a targeted `ComputedExpr`.
+
+In a sense, `ComputedExpr` can be seen as an resolved `Expr` with const variables replaced with actual values. 
+
+### Types of ComputedExpr
+
+```rust
+pub enum ComputedExprKind<F, V>
+where
+    F: BackendField,
+    V: BackendVar,
+{
+    /// Structures behind a custom struct can be recursive, so it embeds the ComputExpr.
+    Struct(BTreeMap<String, ComputedExpr<F, V>>),
+    /// Structures behind an array can be recursive, so it embeds the ComputExpr.
+    Array(Vec<ComputedExpr<F, V>>),
+    Bool(Var<F, V>),
+    Field(Var<F, V>),
+    /// Access to a variable in the scope.
+    Access(Access<F, V>),
+    /// Represents the results of a builtin function call.
+    /// Because we don't know the exact type of the result, we store it as a Var.
+    /// We may deprecate this once it is able to type check the builtin functions,
+    /// so that the result can be inferred.
+    FnCallResult(Var<F, V>),
+}
+```
+
+### Assignment using ComputedExpr
+
+Instead of using `VarOrRef` to narrow down and do the updating of `cvars`, we  use the `ComputedExpr` to represent the access path, which is iteratively determined when stepping through the computations of `Expr::FieldAccess` and `Expr::ArrayAccess`.
+
+When it comes to the computation for `Expr::Assignment`, it can use the access path embeded in a `ComputedExprKind::Access` to locate a nested `ComputedExpr` to swap with a new one. Then it saves the `VarInfo` with same name `houses` as local variable but holding the updated tree of `ComputedExpr` that represents a new set of `cvars`.
+
+### Use ComputedExpr in builtin functions
```

Comment:
so so far:

* I think I understand the usecases this unlock
* I'm not sure why `VarOfRef` is relevant to the explanation, but I understand what it does (maybe the RFC should mention that the main functions in the parser (`compile_stmt` and `compile_expr`) returns a `VarOrRef`
* I don't think the RFC spends enough time explaining why const generics can't work as is currently. IMO it should explain why it doesn't work in two different phases of the compilation: type checker and circuit writer
* this RFC is mostly on functions that take const args, but it doesn't really spell that